### DISCUSSION TITLE
v4.0 burndown list

### DISCUSSION THREAD

**Discussion by RussKie** [gitextensions/gitextensions#9713] [Show and tell] [answered]
This thread's purpose is outline and discuss the features that we'll make the vNext into an official version - v4.0.
The new version will run on **.NET** (currently 5.0, at the time of release will likely be 6.0).

### Release Notes Highlights
* Uses **.NET** 6 #9860 #8522
* Improved repository loadtime considerably (#9243, #9735, #9864 etc) as well as other performance tweaks.
* Use native WSL Git executables for \\wsl$ shares to speedup handling #9702
* Use the standard FormBrowse for FileHistory and blame #9445
* BASE diff: Icons for unique changes (#9720)
* Show stashes in grid (#10138)

### Blocking release:
* ~https://github.com/gitextensions/gitextensions/issues/9402 #9794~
* ~https://github.com/gitextensions/gitextensions/issues/9491 #9794~
* ~https://github.com/gitextensions/gitextensions/issues/9135 #9785~
* ~https://github.com/gitextensions/gitextensions.pluginmanager/pull/58~
* ~https://github.com/gitextensions/gitextensions/pull/9704~
* ~https://github.com/gitextensions/gitextensions/pull/9782~
* ~Some cleanup related to #9445~

### Desirable to include:
~* #9735 *Improves startup and repo switching a lot*~

### Cut:
* PuTTY binaries are no longer bundled with Git Extentions, please install PuTTY separately and change the path in settings.
* Theming is not supported - v3.x relied on EasyHook library to facilitate hooks into Windows API necessary for theming. However .NET 6 does not support remoting and multiple appdomain required for those hooks. To date no viable alternative has been found. https://github.com/gitextensions/gitextensions/pull/9191
* Removed TFS integration, https://github.com/gitextensions/gitextensions/pull/8522/commits/8190d0c0570af95ffbf9e168072d1b169eb99618
* Removed JIRA integration, see: https://github.com/gitextensions/gitextensions/issues/9659

**Comment by gerhardol** [comment]
~#9734 (that requires #9729) is a regression that must be included.~
~#9728 kind of completes features in master~
~#9720 is also highly wanted by me~

~* gitextensions/gitextensions#9702~
~Champion: @gerhardol "_A reason is that I can motivate this at work, checking status etc (not coding or debugging)._"~
~* gitextensions/gitextensions#9386 should be included too.~
~Champion: @gerhardol "_Since we cannot fix that the installer changes the settings._~

**Comment by gerhardol** [comment]
What about .NET6, to be included?

**Comment by RussKie** [comment]
There are blocking issues like #9402.

I have a branch with it, but a) I haven't had a chance to really test it,
and b) AppVeyor didn't natively support .NET 6 (it had to be installed via
scripts). It was a while since I checked last, so things may have changed
since.

**Reply by pmiossec** [reply]
For #9402 and #9491, I think we could do a workaround to not block the release until a fix is released (in .net 7 ?) by removing the panel anchoring.
I have made this workaround in my local version to be able to use it and I can push it if needed.

I can maybe also refine it so that it applies only when the scaling is different than 100% scaling.

And we could imagine removing the workaround when the bug will be fixed...

It's not a perfect workaround but it allows the user to use the push form. With the workaround:

![workaround_frmpush](https://user-images.githubusercontent.com/460196/148101108-69fb08ee-1011-42e6-baaa-5308a4e0316c.gif)

For create branch form:

![workaround_frmcreatebranch](https://user-images.githubusercontent.com/460196/148101495-537d7650-c8dd-4193-92ce-4b58a6c96bb8.gif)

**Reply by gerhardol** [reply]
I believe the time in alpha will need to be longer this time, so this is probably needed.

**Reply by pmiossec** [reply]
I fixed the 2 broken forms but some others doesn't display very well (especially in the settings) even if still usable:

![image](https://user-images.githubusercontent.com/460196/148138865-aaf664a5-0fd1-4caf-855c-18f30a706511.png)

Hidden content:
![image](https://user-images.githubusercontent.com/460196/148138888-0fd89b23-2fac-4023-a26b-19495215512c.png)
![image](https://user-images.githubusercontent.com/460196/148139366-d15efa2e-b703-4ea5-a563-0aeb099256e6.png)


![image](https://user-images.githubusercontent.com/460196/148138907-673e5347-ccdd-4a4a-ba03-cd9ec91196df.png)
![image](https://user-images.githubusercontent.com/460196/148138967-32ed2139-f200-48b8-874f-ddc52860568b.png)
![image](https://user-images.githubusercontent.com/460196/148139041-bfdda07c-22e5-4dc2-979c-7e08bbde045a.png)

And after my PR, I rediscovered [my comment](https://github.com/gitextensions/gitextensions/issues/9402#issuecomment-952630244) saying that the 2 Forms to delete branch and remotes are still broken and I didn't found a fix :( :

![image](https://user-images.githubusercontent.com/460196/148139133-fb84cdbd-eec0-4cf3-80bb-a7a0ea9638ce.png)

edit: I found the issue it is the anchoring `this.branches.Anchor` not in the delete forms, but in `BranchComboBox.Designer.cs`

**Reply by RussKie** [reply]
I cycled through all the Settings pages at 150% and 200% and haven't observed anything that I would consider blocking. There are some visual artifacts, and those could be dealt on ad-hoc basis.

**Comment by gerhardol** [comment]
Issues that should be resolved before a release (not required before an alpha)
* https://github.com/gitextensions/gitextensions/issues/9752
* ~https://github.com/gitextensions/gitextensions/issues/9792~

**Comment by gerhardol** [comment]
A preliminary doc version available here: https://git-extensions-documentation.readthedocs.io/en/feature-4.0-prepare/
PR exists here https://github.com/gitextensions/GitExtensionsDoc/pull/140

**Reply by gerhardol** [reply]
Can someone review the changes?

**Reply by RussKie** [reply]
Merge?

**Reply by gerhardol** [reply]
I will just cleanup some to-do - not really finish but make it look better and build a 4.0

**Comment by RussKie** [comment]
@gerhardol @mstv should we publish a beta? Is there anything truly blocking?

**Reply by gerhardol** [reply]
> @gerhardol @mstv should we publish a beta? Is there anything truly blocking?

Nothing blocking, an alpha should be published.

**Reply by RussKie** [reply]
Here's the first build https://ci.appveyor.com/project/gitextensions/gitextensions/builds/43760552/artifacts

**Comment by RussKie** [comment]
@gerhardol I've updated release/4.0 with all the latest changes from the master. Is there anything else you want to add before we commit to a release?

**Reply by gerhardol** [reply]
Can we include the recent open prs? 
I would like @mstv to review but I can merge tonight.

**Reply by RussKie** [reply]
Sure thing. 
Hopefully @mstv could find some time this week. If not, we can take those and then act on any feedback later.

**Reply by RussKie** [reply]
The latest unsigned build is here: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/44307245/artifacts

**Reply by gerhardol** [reply]
Merged some pending PRs that should be in the beta (changes the  behavior).
There will changes both to behavior and translations, but we need a wider alpha...

**Comment by RussKie** [comment]
I pushed the **v4 alpha1**: https://github.com/gitextensions/gitextensions/releases/tag/v4-alpha1

@gerhardol @mstv @pmiossec please update the release info with anything relevant.

**Comment by gerhardol** [comment]
A few commits with reported issues added to release/4.0, to be included in next build. A few pending PRs in the milestone too.
I would also like to nominate #10138 for inclusion in 4.0.

**Comment by gerhardol** [comment]
@RussKie 
I would like to merge PRs in the following order, most are not reviewed though
All these on both master/4.0, keep the branches identical for code for some time, cherry pick in the same order to 4.0 as in master

* #10140 Regression to the latest commit in master/4.0, nice to keep them togethe
* #10127 Regression in 4.0, related to #10140
* #10154
* #10153
* #10119 Even if not so many has reported this it superannoying when it occurs
* #10117 - or reject
* #10129 Minor regression in 4.0
* #10118 OK for me
* #10145
* #10146
* #10149 Not a really a regression, but exposes this old bug more easily

Milestones are adjusted, set 4.0.0 except for regressions

something I would like to see in 4.0 (I would add it as a feature)
* #10130 (very minor improvement but mostly a simplification)
* #10138 (maybe expose the setting for untracked changes)

* #10172 Decouples and simplifies. As filter handling is changed, this may be needed to make further changes...

**Reply by RussKie** [reply]
Sounds good! No objections from me. Thank you for doing this.
Do you want my help with porting any of these?

**Reply by gerhardol** [reply]
I have submitted most of these prs and would like to have them reviewed. There are a few other 4.0 related issues like deleting branches too. 
As it is now there are no code differences in master and 4.0 branches, only the stash pr that possibly should change that. 

I will make an update to #10129

**Reply by mstv** [reply]
> would like to have them reviewed

I will have a look.

**Reply by RussKie** [reply]
Add to the list
* https://github.com/gitextensions/gitextensions/pull/10159 
* https://github.com/gitextensions/gitextensions/pull/10161

**Reply by gerhardol** [reply]
Time for a new alpha or even beta/RC?

There are some PRs mentioned above that could be considered.
* https://github.com/gitextensions/gitextensions/issues/10122  has been discussed some lately too.

**Reply by RussKie** [reply]
I acknowledge the issue, but it doesn't look to me we're reached a consensus.

**Reply by gerhardol** [reply]
Not for a 'final' solution. To add the force checkbox back handles the regression.
For me, also the regression fix can wait.
Personally, I do not believe that GE should outsmart Git and go back to the old implementation. (I do see a difference between adding functions on top of Git and to out-smart Git, the line is thin though...)

**Reply by pmiossec** [reply]
I would like to see this one land in v4.0:
https://github.com/gitextensions/gitextensions/pull/10183

and maybe https://github.com/gitextensions/gitextensions/pull/10188 because it is a regression (from Git? or GitExtensions ? )

**Reply by gerhardol** [reply]
> I would like to see this one land in v4.0: #10183
> 
> and maybe #10188 because it is a regression (from Git? or GitExtensions ? )

Agree for those two.
What about #10138 ?
Otherwise we should make at least a beta, also with translations.

**Comment by RussKie** [comment]
I'm currently away, I'll be able to produce an installer some time
next week,  if we're ready.

**Comment by gerhardol** [comment]
Milestone 4.0.0 https://github.com/gitextensions/gitextensions/milestone/60
#10193 should be merged before next build
#9223 is added to the milestone too. #10001 at least improves, but tests fail so it is not to be merged.  

Other issues related:
* #10189 to be merged? (Possible followup with adding to Advanced filter, not to delay the release.)
* #10194 not new, but OK fix?
* #10173 could be improved
* #10166 - not fully investigated
* #10165 - add this regression to the milestone, but no remedy
* #10122 Add force checkbox?
* #10009 New in .net6? This applies to all UI elements, a click required to activate window, even if hover etc is working

**Reply by RussKie** [reply]
v4 beta - https://github.com/gitextensions/gitextensions/releases/tag/v4-beta

**Comment by gerhardol** [comment]
Milestone 4.0.0 https://github.com/gitextensions/gitextensions/milestone/60
PRs there should be included in next build
But after next beta, there should be less changes so we can release 4.0

Others
* #10194 I have not got into reproducing the error
* #10250 Set as regression, should be included if the cause is found (mostly important with #10258)
* #10258 Not included in milestone but 'completes' stash-in-grid feature and I do not believe it is a risk to include it

??
* #10009 Settings for filters etc are persisted immediately

**Reply by RussKie** [reply]
I don't think any of these are blockers for the v4 release.

**Comment by RussKie** [comment, marked as answer]
v4 is published: https://github.com/gitextensions/gitextensions/releases/tag/v4
